### PR TITLE
Wav2Vec2 pretraining

### PR DESCRIPTION
**Pull request by ManueleSigona**
# What does this PR do?
Contributors: @callumm-graphcore, @HMellor (and me)
Add Wav2Vec2ForPretraining model in base configuration.
Add example that runs speech pretraining on LibriSpeech dataset.
The model compiles and runs on IPUs.
Tested with SDK 2.5.

## Known issues
The Gumbel vector quantiser uses the function `.exponential_()`; in its current internal PopTorch implementation it draws samples from a uniform distribution [0, 1], of which it then take the `log()`. In the relatively rare cases where a 0 is sampled, it can result in NaNs. We have a suggested solution and task open internally, to address this.

## Known limitation
[Wav2Vec2's pre-training is known to be quite unstable](https://github.com/huggingface/transformers/tree/main/examples/pytorch/speech-pretraining#speech-recognition-pre-training), and aside from the known issue described above, we haven't observed divergence in pretraining. However, we also haven't spent time to verify the convergence, so we cannot guarantee that pretraining convergence is achievable with the current code and default set of hyperparamteres. It might require further work.

### CODE REVIEW

**Comment by jimypbr** [comment]
Please fix the style with `make style`
Add wav2vec2 to the unit tests

File: examples/speech-pretraining/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	--overwrite_output_dir \
+	--layerdrop 0.05 \
+	--per_device_train_batch_size=1 \
+	--dataloader_mode="async_rebatched" \
+	--dataloader_num_workers=8 \
+	--num_train_epochs=10 \
+	--warmup_steps="1000" \
+	--weight_decay="0.01" \
+	--learning_rate="0.001" \
+	--logging_steps="100" \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:100]: Maybe remove the `logging_steps`.

File: examples/speech-pretraining/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	--dataset_name="librispeech_asr" \
+	--dataset_config_name="clean" \
+	--train_split_name="validation" \
+	--ipu_config_name Graphcore/wav2vec2-base-ipu \
+	--output_dir="./wav2vec2-pretrained-demo" \
+	--max_duration_in_seconds="20.0" \
+	--min_duration_in_seconds="2.0" \
+	--do_train \
+	--overwrite_output_dir \
+	--layerdrop 0.05 \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:67]: Is training supported with `layerdrop` > 0 on IPUs?
This seems to be the case (`ipu_layer_drop,py`), but just checking.

File: examples/speech-pretraining/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+---
+
+### Demo
+
+In this demo run we pre-train a `"base-sized"` Wav2Vec2 model simply only on the validation
+data of [librispeech_asr](https://huggingface.co/datasets/librispeech_asr).
+
+```bash
+python run_pretraining.py \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:56]: I think it would be better to follow the same format for the command line example, currently there is a mix of:

- arg_name=arg_value
- arg_name arg_value

File: examples/speech-pretraining/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	--dataset_name="librispeech_asr" \
+	--dataset_config_name="clean" \
+	--train_split_name="train.100" \
+	--ipu_config_name Graphcore/wav2vec2-base-ipu \
+	--output_dir="./wav2vec2-pretrained-base" \
+	--max_duration_in_seconds="20.0" \
+	--min_duration_in_seconds="2.0" \
+	--do_train \
+	--overwrite_output_dir \
+	--layerdrop 0.05 \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:92]: Same question.

File: examples/speech-pretraining/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	--weight_decay="0.01"
+```
+
+### Base
+
+To pre-train `"base-sized"` Wav2Vec2 model, *e.g.* [facebook/wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base)
+on 100h of training data from the [librispeech_asr](https://huggingface.co/datasets/librispeech_asr), the following command can be run:
+
+```bash
+python run_pretraining.py \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:81]: Same comment on the argument format.

File: optimum/graphcore/models/wav2vec2/ipu_gumbel_vector_quantizer.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        ret = y_hard - y_soft.detach() + y_soft
+    else:
+        # Reparametrization trick.
+        ret = y_soft
+
+    return ret
+
+
+class IPUWav2Vec2GumbelVectorQuantizer(Wav2Vec2GumbelVectorQuantizer):
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_gumbel_vector_quantizer.py:51]: Maybe some comments here on why this is needed, this will make things easier for us to track later on.

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+from transformers.models.wav2vec2.modeling_wav2vec2 import (
+    Wav2Vec2Adapter,
+    Wav2Vec2Encoder,
+    Wav2Vec2EncoderStableLayerNorm,
+)
+
+
+class IPUWav2Vec2Encoder(Wav2Vec2Encoder):
+    def __init__(self, config, sequence_length_padding_divisor=1):
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:34]: Why the line break?

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        hidden_states = self.dropout(hidden_states)
+
+        hidden_states = F.pad(
+            hidden_states,
+            # Want e.g. (..., 999, 768) -> (..., 1000, 768)
+            pad=(0, 0, 0, pad_length),
+        )
+
+        for layer in self.layers:
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:94]: Why the line break?

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+            raise ValueError("output_attetntions=True is not supported " "for IPUWav2Vec2EncoderStableLayerNorm")
+        if output_hidden_states:
+            raise ValueError("output_hidden_states=True is not supported " "for IPUWav2Vec2EncoderStableLayerNorm")
+
+        if attention_mask is not None:
+            # make sure padded tokens output 0
+            hidden_states[~attention_mask] = 0.0
+
+            # Pad attention mask to more divisible length
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:145]: Why the line break?

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            remainder = attention_mask.size(-1) % self.sequence_length_padding_divisor
+
+            if remainder != 0:
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:149]: Why the line break?

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+                attention_mask = F.pad(
+                    attention_mask,
+                    # Want e.g. (..., 999) -> (..., 1000)
+                    pad=(0, pad_length),
+                    value=0.0,
+                )
+
+            else:
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:160]: Same thing for the line break.

File: optimum/graphcore/models/wav2vec2/ipu_layer_drop.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        hidden_states = self.dropout(hidden_states)
+
+        hidden_states = F.pad(
+            hidden_states,
+            # Want e.g. (..., 999, 768) -> (..., 1000, 768)
+            pad=(0, 0, 0, pad_length),
+        )
+
+        for layer in self.layers:
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_layer_drop.py:180]: Line break.

File: optimum/graphcore/models/wav2vec2/ipu_wav2vec2_model.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2022 Graphcore Ltd. All rights reserved.
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/ipu_wav2vec2_model.py:1]: I think this can be put in `modeling_wav2vec.py`

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+                encoder = Wav2Vec2Encoder(self.config)
+        else:
+            # Inject IPU Layer Drop
+            # We also pad the sequence length for self-attention
+            # This makes the memory use across tiles more balanced
+            if self.config.do_stable_layer_norm:
+                encoder = IPUWav2Vec2EncoderStableLayerNorm(self.config, sequence_length_padding_divisor=4)
+            else:
+                encoder = IPUWav2Vec2Encoder(self.config, sequence_length_padding_divisor=4)
+        self.wav2vec2.encoder = encoder
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:62]: It seems that you change the instance and not the class, which might cause issues if the layer has a state?

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+            restore: whether to restore the adapter to its original version or not.
+        """
+        if self.config.add_adapter:
+            if restore:
+                adapter = Wav2Vec2Adapter(self.config)
+            else:
+                adapter = IPUWav2Vec2Adapter(self.config)
+        else:
+            adapter = None
+        self.wav2vec2.adapter = adapter
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:77]: Same comment.

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+        """Changes the quantizer class to update its forward pass so that it uses our custom version.
+
+        Args:
+            restore: whether to restore the quantizer to its original version or not.
+        """
+        if restore:
+            quantizer = Wav2Vec2GumbelVectorQuantizer(self.config)
+        else:
+            quantizer = IPUWav2Vec2GumbelVectorQuantizer(self.config)
+        self.quantizer = quantizer
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:89]: Same comment.

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+        ```
+        """
+        super().parallelize()
+
+        self.wav2vec2.__class__ = IPUWav2Vec2Model
+        self.change_wav2vec2_encoder_class(False)
+        self.change_wav2vec2_adapter_class(False)
+        self.change_quantizer_class(False)
+        self.change_conv_eps(False)
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:134]: How does it handle multiple pods since it seems that ipu_ids are hardcoded?

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+        input_values,
+        attention_mask=None,
+        mask_time_indices=None,
+        sampled_negative_indices=None,
+        gumbel_temperature=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:176]: Line break.

**Comment by jimypbr** [comment]
This model is SDK2.5 dependent, correct?

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+        self._add_begin_block(self.wav2vec2.encoder.layers[0], name="EL[00,03)", ipu_id=3)
+
+        self._add_begin_block(self.wav2vec2.encoder.layers[3], name="EL[03,06)", ipu_id=4)
+
+        self._add_begin_block(self.wav2vec2.encoder.layers[6], name="EL[06,09)", ipu_id=5)
+
+        self._add_begin_block(self.wav2vec2.encoder.layers[9], name="EL[09,12)+ELQ", ipu_id=6)
+
+        self._add_begin_block(self.quantizer, name="Quantizer+Losses", ipu_id=7)
```
Comment (jimypbr) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:149]: I think you should specify the block ipus with layers_per_ipu config rather than hard-coding

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+        """
+        super().deparallelize()
+        self.change_wav2vec2_encoder_class(True)
+        self.change_wav2vec2_adapter_class(True)
+        self.change_quantizer_class(True)
+        self.change_conv_eps(True)
+        self.wav2vec2.__class__ = Wav2Vec2Model
+        return self
+
+    def forward(
```
Comment (jimypbr) [inline optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:165]: Is the forward changed at all from the original? Could you just call the `super().forward`?

File: examples/speech-pretraining/README.md (hunk 2 above)
Comment (ManueleSigona) [reply examples/speech-pretraining/README.md:67]: It is; just like you noticed, we implemented an IPU-friendly version in `ipu_layer_drop.py`

File: optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py (hunk 20 above)
Comment (ManueleSigona) [reply optimum/graphcore/models/wav2vec2/modeling_wav2vec2.py:172]: There are a few differences, for example our implementation of the quantizer takes an extra input, so we need to pass it here, then we have some extra padding, some slight differences in some reshapes, etc.

**Comment by jimypbr** [comment]
Can you rebase onto `main` branch please?

File: examples/speech-pretraining/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,6 @@
+transformers==4.18.0
+datasets >= 1.18.0
+torchaudio == 0.10.0
```
Comment (jimypbr) [inline examples/speech-pretraining/requirements.txt:3]: Best use the CPU version of torchaudio.
```
--find-links https://download.pytorch.org/whl/torch_stable.html
torchaudio == 0.10.0+cpu
```

**Comment by jimypbr** [comment]
Rebase this please

File: examples/speech-pretraining/README.md [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+	--model_name_or_path "facebook/wav2vec2-base" \
+	--dataset_name "librispeech_asr" \
+	--dataset_config_name "clean" \
+	--train_split_name "validation" \
+	--ipu_config_name Graphcore/wav2vec2-base-ipu \
+	--output_dir "./wav2vec2-pretrained-demo" \
+	--max_duration_in_seconds "20.0" \
```
Comment (michaelbenayoun) [inline examples/speech-pretraining/README.md:63]: I think we should not put number values under quotes since not all of them are under quotes.
```suggestion
	--max_duration_in_seconds 20.0 \
```

File: setup.py [hunk 23]
```diff
@@ -29,6 +29,8 @@
         "psutil",
         "pytest",
         "pytest-pythonpath",
+        "librosa",
```
Comment (michaelbenayoun) [inline setup.py:32]: I think this should be put in `examples/speech-pretraining/requirements.txt` instead, no?

Comment (ManueleSigona) [reply setup.py:32]: It's there too. But I added it here as well because the test `test_pretrained_and_pipelined_models_match` for Wav2Vec2 in `tests/test_pipelined_models.py` needs this library in order to use speech data

**Review by michaelbenayoun — APPROVED** [review]
Thanks a lot for this HUGE contribution!
Let's check how the CI is doing and we can merge.
